lib/blamer/message.rb

```diff
@@ -0,0 +1,200 @@
+  # fault.
+  #
+  # Matching to a Bug
+  # -----------------
+  #
+  # In order to match an Occurrence to a Bug, the line number of the backtrace
+  # that is at fault must first be determined. This is done probabilistically, by
+  # assigning a score to each line of the backtrace and picking the line with
+  # the highest score. Currently, the factors that influence a line's score are
+  #
+  # * the "height" of that line in the backtrace, and
+  # * the age of the commit that most recently modified that line.
+  #
+  # The highest-scoring file and line number are termed the **relevant file** and
+  # **relevant line**.
+  #
+  # An Occurrence is matched to a Bug using three mandatory criteria:
+  #
+  # * the relevant file path,
+  # * the relevant line number, and
+  # * the class name of the raised exception.
+  #
+  # In addition, if a "blamed" commit is available for the relevant file and line,
+  # the bug is matched on that commit ID as well. Note that the exception message
+  # is _not_ included in the search criteria.
+  #
+  # While each Occurrence maintains the original exception message, the Bug
+  # filters the exception message to remove occurrence-specific information. There
+  # are two ways of filtering exception messages:
+  #
+  # * If the exception can be filtered using {MessageTemplateMatcher}, that
+  #   filtered message is used.
+  # * Otherwise, a general search-and-replace is used to replace likely
+  #   occurrence-specific information (like numbers, inspect values, etc.) with
+  #   placeholders.
+  #
+  # All placeholders take the form "[ALL CAPS IN BRACKETS]" so that the view can
+  # recognize them as such and format them appropriately.
+
+  class Message < Base
+    def find_or_create_bug!
+      
+      criteria = bug_search_criteria
+      bug      = if deploy
+                   # for distributed projects, we need to find either
+                   # * a bug associated with the exact deploy specified in the build attribute
+                   # * or, if the exact deploy specified does not have such a bug:
+                   #   * find any open (not fixed) bug matching the criteria;
+                   #   * if such a bug exists, use that bug and alter its deploy_id to
+                   #     the given deploy;
+                   #   * otherwise, a new bug is created.
+                   b        = Bug.transaction do
+                     environment.bugs.where(criteria.merge(deploy_id: deploy.id)).first ||
+                         environment.bugs.where(criteria.merge(fixed: false)).
+                             find_or_create!(bug_attributes)
+                   end
+                   b.deploy = deploy
+                   b
+                 else
+                   # for hosted projects, we search for any bug matching the criteria under
+                   # our current environment
+                   occurrence.message = occurrence.message.partition(/[^"]+/)[1]
```

Comment:
Could you add a comment explaining your regex and some example patterns that you expect to match?